File: app/src/main/java/org/ligi/ipfsdroid/IPFSDaemonService.java
package org.ligi.ipfsdroid;

import android.app.IntentService;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.support.v4.app.NotificationCompat;

public class IPFSDaemonService extends IntentService {

    private NotificationManager nManager;
    int NOTIFICATION_ID = 12345;

    public IPFSDaemonService() {
        super("IPFSDaemonService");
    }

    @Override
    protected void onHandleIntent(final Intent intent) {
        Intent exitIntent = new Intent(this, IPFSDaemonService.class);
        exitIntent.setAction("STOP");
        PendingIntent pendingExit = PendingIntent.getService(this, 0, exitIntent, 0);

        Intent targetIntent = new Intent(this, DetailsActivity.class);
        PendingIntent pIntent = PendingIntent.getActivity(this, 0, targetIntent, 0);
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this).setOngoing(true)
                                                                                 .setSmallIcon(R.drawable.notification)
                                                                                 .setContentTitle("IPFS Daemon")
                                                                                 //.addAction(R.drawable.ic_navigation_check, "exit", pendingExit)
                                                                                 .setContentText("The daemon is running");


        builder.setContentIntent(pIntent);
        nManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        nManager.notify(NOTIFICATION_ID, builder.build());

        try {
            new IPFSDaemon(getBaseContext()).run("daemon").waitFor();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }


    @Override
    public void onDestroy() {
        super.onDestroy();
        if (nManager!=null) {
            nManager.cancel(NOTIFICATION_ID);
        }

    }


    @Override
    public int onStartCommand(final Intent intent, final int flags, final int startId) {
        final String action = intent.getAction();
        if (nManager!=null && action != null && action.equals("STOP")) {
            // TODO actually stop the daemon https://github.com/ipfs/faq/issues/39
        }
        return super.onStartCommand(intent, flags, startId);
    }

}

File: app/src/main/java/org/ligi/ipfsdroid/IPFSDaemonService.kt
package org.ligi.ipfsdroid

import android.app.IntentService
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.support.v4.app.NotificationCompat
import org.ligi.ipfsdroid.activities.DetailsActivity

class IPFSDaemonService : IntentService("IPFSDaemonService") {

    private var nManager: NotificationManager? = null
    private var daemon: Process? = null
    internal var NOTIFICATION_ID = 12345

    override fun onHandleIntent(intent: Intent) {
        val exitIntent = Intent(this, IPFSDaemonService::class.java)
        exitIntent.action = "STOP"
        val pendingExit = PendingIntent.getService(this, 0, exitIntent, 0)

        val targetIntent = Intent(this, DetailsActivity::class.java)
        val pIntent = PendingIntent.getActivity(this, 0, targetIntent, 0)
        val builder = NotificationCompat.Builder(this).setOngoing(true)
                .setSmallIcon(R.drawable.notification)
                .setContentTitle("IPFS Daemon")
                .setContentText("The daemon is running")
                .addAction(android.R.drawable.ic_menu_close_clear_cancel, "exit", pendingExit)

        builder.setContentIntent(pIntent)
        nManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        nManager!!.notify(NOTIFICATION_ID, builder.build())

        try {
            daemon = IPFSDaemon(baseContext).run("daemon")
            State.isDaemonRunning = true
            daemon!!.waitFor()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    override fun onDestroy() {
        daemon!!.destroy()
        super.onDestroy()
        State.isDaemonRunning = false
        nManager?.cancel(NOTIFICATION_ID)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val action = intent.action
        if (nManager != null && action != null && action == "STOP") {
            stopSelf()
        }
        return super.onStartCommand(intent, flags, startId)
    }

}
